core/ClassInfo.php

```diff
@@ -52,17 +52,20 @@ public static function hasTable($class) {
 			return false;
 		}
 	}
-	
+
 	public static function reset_db_cache() {
 		self::$_cache_all_tables = null;
 		self::$_cache_ancestry = array();
 	}
-	
+
 	/**
 	 * Returns the manifest of all classes which are present in the database.
 	 * @param string $class Class name to check enum values for ClassName field
 	 */
 	public static function getValidSubClasses($class = 'SiteTree', $includeUnbacked = false) {
+		if (self::hasTable($class)) {
+			$class = self::$_cache_all_tables[strtolower($class)];
```

Comment:
This seems complicated to me. singleton($class)->class is a lot less work, and DataObjects can't be abstract anyway so no chance of that breaking here.